Implement a component dropdown

weierophinney opened the pull request:
This patch implements a component-dropdown.

The "dropdown" markup is placed below the footer, with a CSS rule to hide it.

~~On page load, it moves the dropdown element to the top of the page, and updates the navbar site logo link to toggle it, additionally adding a tooltip that indicates clicking the logo will toggle the list.~~

On first click of the "ZF" logo, it retrieves the list of component documentation from https://zendframework.github.io/zf-mkdoc-theme/scripts/zf-component-list.json and injects entries for each item into the DOM; once complete, it displays the list. (Subsequent displays do not need to re-retrieve the list, for obvious reasons.)

An additional link is added at the bottom of the markup to hide the list; hitting the logo again will also hide it.

The functionality works on both mobile and desktop versions. I've pushed a build of it to the [zend-stdlib](https://zendframework.github.io/zend-stdlib/) docs to demonstrate usage.

weierophinney commented:
WIP; currently also experimenting with a JSONP approach, which would allow us to update the list independent of the individual component documentation pages.

weierophinney commented:
Ready for review!

froschdesign commented:
I will check this.

froschdesign commented on theme/scripts.html:
```diff
@@ -1,48 +1,90 @@
 <script language="javascript" type="text/javascript">
```
Remove `language` and `type` attributes. [`type` is optional](https://developer.mozilla.org/en/docs/Web/HTML/Element/script#attr-type) and `language` was already deprecated in HTML4.

froschdesign commented on theme/dropdown.html:
```diff
@@ -1,184 +1,11 @@
-    
+  <div class="zf-components-grid">
     <div class="zf-logo">
       <a href="http://framework.zend.com" data-toggle="tooltip" data-placement="bottom" title="Go to the ZF homepage""><img src="http://framework.zend.com/images/logos/zf-logo-mark.png" alt="Zend Framework" width="250" /></a>
```
After the `title` attribute there are two quotes. Remove one.

froschdesign commented:
The JSON list is a nice idea.

froschdesign commented:
Click on the logo? Who knows this? There is a tooltip, but I have to "rollover" the logo to see the hint.
If I click on the logo, I would expect to return to the start page.

My proposal:

- A general top bar with a logo, component-dropdown and GitHub-link.
- A second bar with name, links and search for the current component.

The component-dropdown could be smaller. I think, a custom drop down with description would be fine.

Like this: 
![dropdown](https://cloud.githubusercontent.com/assets/103362/14309997/f45d64b8-fbdf-11e5-8262-d2949d8b3659.png)

basz commented:
I agree with froschdesign that discovering this menu by clicking the logo is not very obvious UX. A similar button as the search, next, prev would be my initial choice. A simple show components button.

I have less problem with the way the menu slides from the top. Having the top-bar moving right from under your mouse/finger isn't optimal. Clicking on the logo (the big one) doesn't 'closes' but will make you leave. Needs some getting used to, but is ok.

I know how easy it is to critique work, so please don't let this discourage you. Thanks for the work, it's a much needed feature!

froschdesign commented:
@basz 
> A similar button as the search, next, prev would be my initial choice.

I would go one step further and remove the next and previous buttons from the top bar. Add these buttons above the content section. Below the content there are already these two buttons.
The same like in the old documentation:

![docs](https://cloud.githubusercontent.com/assets/103362/14315834/e86349dc-fbff-11e5-848b-840ebb121cdd.png)

> …so please don't let this discourage you.

Nobody wants this! Only fair comment.

weierophinney commented:
Thanks for the great feedback!

I was indeed hesitant about re-using the ZF logo for this, for the reasons @froschdesign outlined: discoverability. However, with the number of items already in the navbar, I wasn't sure exactly how to proceed. I'll investigate pushing the prev/next links into the content, and replacing them with a link to expose the component list; my main worry about this is that having them in the navbar meant they were visible/reachable always, no matter where in the document you were.

With regards to doing a dropdown select, this is a no-go from the outset. The list of components currently is reasonably manageable, but as we get more and more components with documentation in place, such a list will become as unwieldy as the one for selecting changelog versions on the current ZF website (we've had *many* complaints about that one!). Additionally, readability becomes a concern; it's useful to have the "tagline" description for each component, and if this text gets bumped too small, it becomes too hard to read. (And zooming will make the dropdown look pretty nasty!)

I'll try another iteration, and drop a note when it's ready for review.

basz commented:
> but as we get more and more components with documentation in place, such a list will become as unwieldy 

Did not think of that. I actually like that it is free form. It resembles a list, but can be anything if needed. (latest component versioning number, open issues, etc... might be of interest)

Would it be possible to change it such that instead of opening above the nav-bar it would appear between the nav-bar and the main content? That way focus isn't lost so much...

Finally perhaps some css tweaking, as the animation is a little bit jagged... (os x safari) A `transform: translateZ(0);` on the div.zf-components smooths things out.

froschdesign commented:
> but as we get more and more components with documentation in place, such a list will become as unwieldy

Extend the list: (example)

![dropdown](https://cloud.githubusercontent.com/assets/103362/14323677/c8346214-fc21-11e5-9389-c7a1e1319d8b.png)

weierophinney commented:
> Would it be possible to change it such that instead of opening above the nav-bar it would appear between the nav-bar and the main content? That way focus isn't lost so much...

I tried it, but was unhappy with the results.

At this time, I've pushed a new commit to this repo, and updated the zend-stdlib docs to demonstrate. The specific changes I made were:

- Moved the prev/next links from the topnav and into the sidenav. This makes them ever-present on screens of 1024px+, and essentially creates top/bottom nav for smaller devices.
- Created a new "component" icon+link, as a peer to the search and github items in the topnav. This is the toggle for the component list now, instead of the ZF logo.
- Added some logic to the toggling to change the sidenav position to "relative" when toggling the component list, and back to its original value when closed again. This prevents it from displaying over the component list.
- Added the translateZ transform to the zf-components element (thanks, @basz!)
- Removed the attributes from the script tag.
- Removed the extra quote in the "hide list" link attributes.
